internal/report/table/table.go

```diff
@@ -0,0 +1,144 @@
+package table
+
+import (
+	"fmt"
+	"github.com/fatih/color"
+	"github.com/jedib0t/go-pretty/v6/table"
+	"github.com/jedib0t/go-pretty/v6/text"
+	"github.com/saucelabs/saucectl/internal/report"
+	"io"
+	"sync"
+	"time"
+)
+
+var defaultTableStyle = table.Style{
+	Name: "saucy",
+	Box: table.BoxStyle{
+		BottomLeft:       "└",
+		BottomRight:      "┘",
+		BottomSeparator:  "",
+		EmptySeparator:   text.RepeatAndTrim(" ", text.RuneCount("+")),
+		Left:             "│",
+		LeftSeparator:    "",
+		MiddleHorizontal: "─",
+		MiddleSeparator:  "",
+		MiddleVertical:   "",
+		PaddingLeft:      "  ",
+		PaddingRight:     "  ",
+		PageSeparator:    "\n",
+		Right:            "│",
+		RightSeparator:   "",
+		TopLeft:          "┌",
+		TopRight:         "┐",
+		TopSeparator:     "",
+		UnfinishedRow:    " ...",
+	},
+	Color: table.ColorOptionsDefault,
+	Format: table.FormatOptions{
+		Footer: text.FormatDefault,
+		Header: text.FormatDefault,
+		Row:    text.FormatDefault,
+	},
+	HTML: table.DefaultHTMLOptions,
+	Options: table.Options{
+		DrawBorder:      false,
+		SeparateColumns: false,
+		SeparateFooter:  true,
+		SeparateHeader:  true,
+		SeparateRows:    false,
+	},
+	Title: table.TitleOptionsDefault,
+}
+
+// Reporter is a table writer implementation for report.Reporter.
+type Reporter struct {
+	TestResults []report.TestResult
+	Dst         io.Writer
+	lock        sync.Mutex
+}
+
+// Add adds the rest result to the summary table.
+func (r *Reporter) Add(t report.TestResult) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.TestResults = append(r.TestResults, t)
+}
+
+// Render renders out a test summary table to the destination of Reporter.Dst.
+func (r *Reporter) Render() {
+	r.lock.Lock()
```

Comment:
Just a question for my own understanding of concurrency in go. Why does `Render` need the lock? Isn't `Render` guaranteed to be called after all the results are collected?

Reply:
> Isn't Render guaranteed to be called after all the results are collected?

Yesno :D That's the way it _ought to be_ and is implemented as such. But from the implementation perspective, the lock is the only way to enforce correct behavior even if the caller were to call the functions incorrectly.